File: src/typescript/helpers/data-structures/graph.ts
export class Graph<T> {
  private neighborsByVertex = new Map<T, T[]>();

  addVertex(vertex: T): void {
    if (!this.neighborsByVertex.has(vertex)) {
      this.neighborsByVertex.set(vertex, []);
    }
  }

  addEdge(a: T, b: T): void {
    this.addVertex(a);
    this.addVertex(b);
    this.neighborsByVertex.get(a)!.push(b);
    this.neighborsByVertex.get(b)!.push(a);
  }

  bfs(start: T): T[] {
    if (!this.neighborsByVertex.has(start)) return [];

    const visited = new Set<T>([start]);
    const order: T[] = [];

    const queue: T[] = [start];
    let headIndex = 0;

    while (headIndex < queue.length) {
      const current = queue[headIndex++];
      order.push(current);

      for (const neighbor of this.neighbors(current)) {
        if (!visited.has(neighbor)) {
          visited.add(neighbor);
          queue.push(neighbor);
        }
      }
    }

    return order;
  }

  dfsIterative(start: T): T[] {
    if (!this.neighborsByVertex.has(start)) return [];

    const visited = new Set<T>();
    const order: T[] = [];
    const stack: T[] = [start];

    while (stack.length > 0) {
      const current = stack.pop()!;

      if (visited.has(current)) continue;
      visited.add(current);
      order.push(current);

      for (const neighbor of this.neighbors(current)) {
        if (!visited.has(neighbor)) {
          stack.push(neighbor);
        }
      }
    }

    return order;
  }

  dfsRecursive(start: T): T[] {
    if (!this.neighborsByVertex.has(start)) return [];

    const visited = new Set<T>();
    const order: T[] = [];

    const visit = (current: T): void => {
      visited.add(current);
      order.push(current);

      for (const neighbor of this.neighbors(current)) {
        if (!visited.has(neighbor)) {
          visit(neighbor);
        }
      }
    };

    visit(start);
    return order;
  }

  hasPath(start: T, end: T): boolean {
    return this.findPath(start, end) !== null;
  }

  /**
   * Finds a path using BFS.
   * In an unweighted graph, BFS also gives the shortest path (fewest edges).
   */
  findPath(start: T, end: T): T[] | null {
    return this.shortestPath(start, end);
  }

  /**
   * Shortest path in an unweighted graph (BFS).
   */
  shortestPath(start: T, end: T): T[] | null {
    if (start === end)
      return this.neighborsByVertex.has(start) ? [start] : null;
    if (!this.neighborsByVertex.has(start) || !this.neighborsByVertex.has(end))
      return null;

    const visited = new Set<T>([start]);
    const parentByVertex = new Map<T, T>();

    const queue: T[] = [start];
    let headIndex = 0;

    while (headIndex < queue.length) {
      const current = queue[headIndex++];

      for (const neighbor of this.neighbors(current)) {
        if (visited.has(neighbor)) continue;

        visited.add(neighbor);
        parentByVertex.set(neighbor, current);

        if (neighbor === end) {
          return this.reconstructPath(parentByVertex, start, end);
        }

        queue.push(neighbor);
      }
    }

    return null;
  }

  private neighbors(vertex: T): T[] {
    return this.neighborsByVertex.get(vertex) ?? [];
  }

  private reconstructPath(parentByVertex: Map<T, T>, start: T, end: T): T[] {
    const path: T[] = [end];
    let current = end;

    while (current !== start) {
      const parent = parentByVertex.get(current);
      if (parent === undefined) return null as any; // should not happen if used correctly
      current = parent;
      path.unshift(current);
    }

    return path;
  }
}

File: src/go/helpers/data-structures/graph.go
package datastructures

type Graph struct {
	adjacencyList map[string][]string
}

func NewGraph() *Graph {
	return &Graph{
		adjacencyList: make(map[string][]string),
	}
}

func (graph *Graph) AddVertex(vertex string) {
	_, exists := graph.adjacencyList[vertex]
	if !exists {
		graph.adjacencyList[vertex] = []string{}
	}
}

func (graph *Graph) AddEdge(firstVertex string, secondVertex string) {
	graph.AddVertex(firstVertex)
	graph.AddVertex(secondVertex)

	graph.adjacencyList[firstVertex] = append(graph.adjacencyList[firstVertex], secondVertex)
	graph.adjacencyList[secondVertex] = append(graph.adjacencyList[secondVertex], firstVertex)
}

func (graph *Graph) BFS(startVertex string) []string {
	_, exists := graph.adjacencyList[startVertex]
	if !exists {
		return []string{}
	}

	visited := make(map[string]bool)
	result := make([]string, 0)

	queue := make([]string, 0)
	queue = append(queue, startVertex)
	visited[startVertex] = true

	queueIndex := 0
	for queueIndex < len(queue) {
		currentVertex := queue[queueIndex]
		queueIndex++

		result = append(result, currentVertex)

		for _, neighbor := range graph.adjacencyList[currentVertex] {
			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, neighbor)
			}
		}
	}

	return result
}

func (graph *Graph) DFSIterative(startVertex string) []string {
	_, exists := graph.adjacencyList[startVertex]
	if !exists {
		return []string{}
	}

	visited := make(map[string]bool)
	result := make([]string, 0)

	stack := make([]string, 0)
	stack = append(stack, startVertex)

	for len(stack) > 0 {
		lastIndex := len(stack) - 1
		currentVertex := stack[lastIndex]
		stack = stack[:lastIndex]

		if visited[currentVertex] {
			continue
		}

		visited[currentVertex] = true
		result = append(result, currentVertex)

		for _, neighbor := range graph.adjacencyList[currentVertex] {
			if !visited[neighbor] {
				stack = append(stack, neighbor)
			}
		}
	}

	return result
}

func (graph *Graph) DFSRecursive(startVertex string) []string {
	_, exists := graph.adjacencyList[startVertex]
	if !exists {
		return []string{}
	}

	visited := make(map[string]bool)
	result := make([]string, 0)

	graph.dfsRecursive(startVertex, visited, &result)

	return result
}

func (graph *Graph) dfsRecursive(
	currentVertex string,
	visited map[string]bool,
	result *[]string,
) {
	visited[currentVertex] = true
	*result = append(*result, currentVertex)

	for _, neighbor := range graph.adjacencyList[currentVertex] {
		if !visited[neighbor] {
			graph.dfsRecursive(neighbor, visited, result)
		}
	}
}

func (graph *Graph) HasPath(startVertex string, endVertex string) bool {
	if startVertex == endVertex {
		return true
	}

	_, startExists := graph.adjacencyList[startVertex]
	_, endExists := graph.adjacencyList[endVertex]
	if !startExists || !endExists {
		return false
	}

	visited := make(map[string]bool)
	queue := make([]string, 0)
	queue = append(queue, startVertex)
	visited[startVertex] = true

	queueIndex := 0
	for queueIndex < len(queue) {
		currentVertex := queue[queueIndex]
		queueIndex++

		for _, neighbor := range graph.adjacencyList[currentVertex] {
			if neighbor == endVertex {
				return true
			}

			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, neighbor)
			}
		}
	}

	return false
}

func (graph *Graph) FindPath(startVertex string, endVertex string) []string {
	// Returns any path (not necessarily shortest). Implemented with BFS.
	// For shortest path use ShortestPath.
	if startVertex == endVertex {
		return []string{startVertex}
	}

	_, startExists := graph.adjacencyList[startVertex]
	_, endExists := graph.adjacencyList[endVertex]
	if !startExists || !endExists {
		return nil
	}

	visited := make(map[string]bool)
	parent := make(map[string]string)

	queue := make([]string, 0)
	queue = append(queue, startVertex)
	visited[startVertex] = true

	queueIndex := 0
	for queueIndex < len(queue) {
		currentVertex := queue[queueIndex]
		queueIndex++

		for _, neighbor := range graph.adjacencyList[currentVertex] {
			if visited[neighbor] {
				continue
			}

			visited[neighbor] = true
			parent[neighbor] = currentVertex

			if neighbor == endVertex {
				return reconstructPath(parent, startVertex, endVertex)
			}

			queue = append(queue, neighbor)
		}
	}

	return nil
}

func (graph *Graph) ShortestPath(startVertex string, endVertex string) []string {
	// BFS on an unweighted graph yields the shortest path in number of edges.
	if startVertex == endVertex {
		return []string{startVertex}
	}

	_, startExists := graph.adjacencyList[startVertex]
	_, endExists := graph.adjacencyList[endVertex]
	if !startExists || !endExists {
		return nil
	}

	visited := make(map[string]bool)
	parent := make(map[string]string)

	queue := make([]string, 0)
	queue = append(queue, startVertex)
	visited[startVertex] = true

	queueIndex := 0
	for queueIndex < len(queue) {
		currentVertex := queue[queueIndex]
		queueIndex++

		for _, neighbor := range graph.adjacencyList[currentVertex] {
			if visited[neighbor] {
				continue
			}

			visited[neighbor] = true
			parent[neighbor] = currentVertex

			if neighbor == endVertex {
				return reconstructPath(parent, startVertex, endVertex)
			}

			queue = append(queue, neighbor)
		}
	}

	return nil
}

func reconstructPath(parent map[string]string, startVertex string, endVertex string) []string {
	path := make([]string, 0)

	currentVertex := endVertex
	path = append(path, currentVertex)

	for currentVertex != startVertex {
		nextVertex, exists := parent[currentVertex]
		if !exists {
			return nil
		}

		currentVertex = nextVertex
		path = append(path, currentVertex)
	}

	reverseStringsInPlace(path)
	return path
}

func reverseStringsInPlace(values []string) {
	leftIndex := 0
	rightIndex := len(values) - 1

	for leftIndex < rightIndex {
		values[leftIndex], values[rightIndex] = values[rightIndex], values[leftIndex]
		leftIndex++
		rightIndex--
	}
}
